### DISCUSSION TITLE
Programmatically add new node connected to existing node

### DISCUSSION THREAD

**Discussion by stanislaw-grin** [xyflow/xyflow#2866] [General]
Hello community,

I'm building a flow where users can dynamically add new nodes connected to the existing one (by clicking on a button inside node and selecting a type of the next node). Each node can have multiple unlimited incomers and outcomes. The behavior I'd like to achieve is the same as in make.com (former Integromat).
My questions are:
1. If nodes can have multiple incomers and outcomes that are created dynamically, should I have handles for each node, or is it enough to have only two - source and target? 
2. How to render Handles dynamically with appropriate handle IDs in order to be able to save & restore built flow? I can't hard-code handles IDs since nodes will be added dynamically, so I use a memorized function that returns new ID on first render, but how to restore actual handles IDs after restoring the flow? Custom node component knows nothing about its handles.
3. I see a [useUpdateNodeInternals](https://reactflow.dev/docs/api/hooks/use-update-node-internals/) hook described in the documentation, but it's not quite clear how to use it. Are there any examples of using it with dynamic changing of handles number? 
4. Also, is it possible to pass custom props to a custom node component? For example, I'd like to pass a callback that will be invoked when some button is clicked inside a node. Was looking for a such question in issues and discussions but haven't found it.

Sorry if some questions are silly. 

Thanks!

**Comment by stanislaw-grin** [comment]
I am shooting all questions, except for the last one about custom props :) I managed to deal with the rest.

**Reply by moklick** [reply]
Great that you could already solve your questions :) You can pass custom functions and stuff inside the data object.

**Reply by shree1999** [reply]
Hi @stanislav-grin I know it's quite late can you share your work here? It would be highly helpful if you can provide details on how you achieved the following 

> by clicking on a button inside node and selecting a type of the next node

how were you able to add nodes dynamically
